File: SeleniumScreenshotCapture/src/main/java/SeleniumScreenshotCapture.java
import org.openqa.selenium.*;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import ru.yandex.qatools.ashot.AShot;
import ru.yandex.qatools.ashot.Screenshot;
import ru.yandex.qatools.ashot.shooting.ShootingStrategies;

import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;


public class SeleniumScreenshotCapture {

	private static void createThumbnail(File input, String output) {
		try {
			BufferedImage sourceImage = ImageIO.read(input);
			BufferedImage image = new BufferedImage(100, 100, BufferedImage.TYPE_INT_RGB);
			Image scaledImage = sourceImage.getScaledInstance(100, 100, Image.SCALE_SMOOTH);
			image.createGraphics().drawImage(scaledImage, 0, 0, null);
			ImageIO.write(image, "jpg", new File(output));
		} catch (Exception e) {
			System.out.println("Unable to create thumbnail: " + e);
		}
	}

	private static void waitForFile(File filepath) {
		try {
			for (int i = 0; i < 5; i++) {
				if (!filepath.exists()) {
					System.out.println("Waiting for screenshot file " + filepath.getName() + " to generate...");
					Thread.sleep(10000);
				}
			}
		} catch (Exception e) {

		}
	}

	// Generate fullpage, screen-sized, and thumbnail screenshots for a given url
	// Should always have an argument for the url=%url% filepath=%image.jpg% output filepath
	// For screen screenshot, inlcude width=val height=val arguments e.g. width=1400 height=800
	// For wayback fullpage screenshot, include --wayback argument
	public static void main(String[] args) throws IOException, InterruptedException {
		String url = null;
		String filepath = null;
		String imageWidth = null;
		String imageHeight = null;
		boolean isWayback = false;

		// Assign variable values based on arguments
		for (String arg : args) {
			String[] keyValues = arg.split("=");
			if (keyValues[0].equals("url")) url = keyValues[1];
			else if (keyValues[0].equals("filepath")) filepath = keyValues[1];
			else if (keyValues[0].equals("width")) imageWidth = keyValues[1];
			else if (keyValues[0].equals("height")) imageHeight = keyValues[1];
			else if (keyValues[0].equals("--wayback")) isWayback = true;
			else {
				System.out.println("Unrecognised argument '" + arg + "'.  Cannot generate screenshot.");
				System.exit(1);
			}
		}

		try {
			// Prepare the driver
			String chromeDriver = Paths.get("chromedriver").toFile().getAbsolutePath();

			System.setProperty("webdriver.chrome.driver", chromeDriver);

			ChromeOptions chromeOptions = new ChromeOptions();
			chromeOptions.addArguments("--headless");
			chromeOptions.addArguments("--no-sandbox");

			if (imageHeight != null && imageWidth != null) {
				chromeOptions.addArguments("--window-size=" + imageWidth + "," + imageHeight);
			}

			WebDriver driver = new ChromeDriver(chromeOptions);

			if (imageWidth == null && imageHeight == null) {
				driver.manage().window().maximize();
			}

			driver.get(url);

			// Change the focus out of the wayback banner
			if (isWayback) {
				// Remove pywb banner
				try {
					JavascriptExecutor jsExecutor = (JavascriptExecutor) driver;
					WebElement banner = driver.findElement(By.id("_wb_frame_top_banner"));
					jsExecutor.executeScript("arguments[0].parentNode.removeChild(arguments[0])", banner);

					// Modify iframe padding
					WebElement frame = driver.findElement(By.id("wb_iframe_div"));
					jsExecutor.executeScript("arguments[0].setAttribute('style', 'padding:0px 0px 0px 0px')", frame);
				} catch (Exception e) {
				}

				WebDriverWait wait = new WebDriverWait(driver, 4000);
				wait.until(ExpectedConditions.visibilityOfElementLocated((By.id("replay_iframe"))));
				driver.switchTo().frame("replay_iframe");
			}

			// Generate the screenshot
			if (imageWidth != null && imageHeight != null) {
				File newFile = ((TakesScreenshot) driver).getScreenshotAs(OutputType.FILE);
				File outputFile = new File(filepath);
				Files.copy(newFile.toPath(), Paths.get(filepath), StandardCopyOption.REPLACE_EXISTING);
				waitForFile(outputFile);

				// for fullpage
			} else {
				Screenshot fullScreen = new AShot().shootingStrategy(ShootingStrategies.viewportPasting(1000)).takeScreenshot(driver);
				ImageIO.write(fullScreen.getImage(), "jpg", new File(filepath));
				waitForFile(new File(filepath));
			}
			driver.quit();

			// Generate a thumbnail for the screenshot
			String size = "";
			if (filepath.contains("fullpage")) {
				size = "fullpage";
			} else if (filepath.contains("screen")) {
				size = "screen";
			}
			createThumbnail(new File(filepath), filepath.replace(size, size + "-thumbnail"));

			System.exit(0);
		} catch (Exception e) {
			System.out.println("Unable to capture the screenshot.  " + e.getMessage());
		}
	}
}

File: SeleniumScreenshotCapture/SeleniumScreenshotCapture.py
#!/usr/bin/python3

import os
import sys

from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions
from PIL import Image


def create_thumbnail(input, output):
    source_image = Image.open(input)
    resized = source_image.resize((200, 150))
    resized.save(output)
    source_image.close()


# Main
def main(command_args):
    url = None
    filepath = None
    width = None
    height = None
    server = None
    is_wayback = False
    wayback_type = None
    wayback_version = None

    # Assign argument values to variables
    for arg in command_args:
        if arg == "--wayback":
            is_wayback = True
            continue
        if "SeleniumScreenshotCapture.py" in arg:
            continue
        key_vals = arg.split("=")
        if key_vals[0] == "url":
            url = key_vals[1]
        elif key_vals[0] == "filepath":
            filepath = key_vals[1]
        elif key_vals[0] == "width":
            width = key_vals[1]
        elif key_vals[0] == "height":
            height = key_vals[1]
        elif key_vals[0] == "selenium-server":
            server = key_vals[1]
        elif key_vals[0] == "wayback-name":
            wayback_type = key_vals[1]
        elif key_vals[0] == "wayback-version":
            wayback_version = key_vals[1]
        else:
            print("Warning: unrecognised argument.  Arg: " + arg)
            # sys.exit(1)

    # file_directory = Path(filepath).parent
    # file_directory.mkdir(parents=True, exist_ok=True)

    # Set up the web driver
    chrome_options = webdriver.ChromeOptions()
    chrome_options.add_argument("--headless")

    if server is not None:
        driver = webdriver.Remote(command_executor=server, options=chrome_options)
    else:
        driver = webdriver.Chrome(options=chrome_options)

    if is_wayback:
        if wayback_type is None:
            wayback_type = "pywb"
        else:
            wayback_type = str(wayback_type).lower()
        if wayback_version is None:
            wayback_version = "2.7.3"
        else:
            wayback_version = str(wayback_version)

        if wayback_type == "owb":
            driver.get(url)
            # Remove wayback banner and modify the iframe
            wait = WebDriverWait(driver, 10)
            wait.until(expected_conditions.visibility_of_element_located((By.ID, "wm-ipp")))
            driver.execute_script("return document.getElementById('wm-ipp').style.display='none';")
        elif wayback_type == "pywb":
            iframe_style = "display:block; overflow:hidden; height:100%; width:100%; margin:0px; border:0px; padding:0px;"
            iframe_html = f"<iframe id='replay_iframe' src='{url}' frameborder='0' style='{iframe_style}'></iframe>"
            html = f"""<!DOCTYPE html>
                    <html style='width: 100wh; height: 100vh; margin:0px; border:0px; padding:0px; overflow: hidden;'>
                    <body style='width: 100wh; height: 100vh; margin:0px; border:0px; padding:0px; overflow: hidden;'>
                    {iframe_html}
                    </body>
                    </html>"""
            driver.get(f"data:text/html;charset=utf-8,{html}")
            wait = WebDriverWait(driver, 10)
            wait.until(expected_conditions.visibility_of_element_located((By.ID, "replay_iframe")))
            driver.switch_to.frame("replay_iframe")
        else:
            print(f"Not supported.")
            sys.exit(-1)
    else:
        driver.get(url)

    driver.execute_script("window.scrollTo(0, 0)")

    # Set screenshot size
    if width is None and height is None:
        driver.fullscreen_window()
        s = lambda x: driver.execute_script('return document.body.parentNode.scroll' + x)
        driver.set_window_size(s('Width') + 20, s('Height'))
    else:
        driver.set_window_size(width, height)

    driver.save_screenshot(filepath)
    driver.close()

    # Generate screenshot thumbnails
    size = None
    if "screen" in filepath:
        size = "screen"
    elif "fullpage" in filepath:
        size = "fullpage"

    create_thumbnail(filepath, filepath.replace(size, size + "-thumbnail"))

    # sys.exit(0)


if __name__ == "__main__":
    # Export Environment
    curr_directory = os.path.dirname(__file__)
    path = os.environ.get("PATH")
    if path is None:
        path = curr_directory
    elif curr_directory not in path:
        path = path + ":" + curr_directory
    os.environ["PATH"] = path
    main(sys.argv)
